File: Real-Life/go/variable.go
package main

import (
    "fmt"
    "math"
    "strconv"
)

/*
 * Variables in Go: A Detailed Explanation
 *
 * A variable is a named storage location in memory that holds a value.
 * Variables allow programs to store, retrieve, and manipulate data.
 *
 * Key Points:
 * 1. Go is statically typed - variables have a specific type
 * 2. Go supports type inference - the compiler can deduce types
 * 3. Variables must be used once declared (unused variables cause compile errors)
 * 4. Go has unique declaration syntax options
 * 5. Variables have a scope that determines where they can be accessed
 *
 * Basic Syntax Options:
 * - var name type = value   (standard declaration)
 * - var name type           (declaration with zero value)
 * - var name = value        (type inference)
 * - name := value           (short declaration with type inference)
 *
 * Common Data Types in Go:
 * - bool: true or false
 * - string: text in double quotes
 * - int, int8, int16, int32, int64: integers with varying sizes
 * - uint, uint8, uint16, uint32, uint64: unsigned integers
 * - float32, float64: floating-point numbers
 * - complex64, complex128: complex numbers
 * - byte: alias for uint8
 * - rune: alias for int32 (represents a Unicode code point)
 */

func main() {
    // Example 1: Basic Variable Declaration and Initialization
    var age int = 30
    fmt.Println("Age:", age)

    // Example 2: Type Inference - compiler determines the type
    var name = "John Doe"
    fmt.Printf("Name: %s (Type: %T)\n", name, name)

    // Example 3: Short Variable Declaration (most common in functions)
    salary := 75000.50
    fmt.Printf("Salary: $%.2f (Type: %T)\n", salary, salary)

    // Example 4: Zero Values (default values when no initial value provided)
    var initialized int = 100
    var uninitialized int // Will be 0
    fmt.Println("Initialized:", initialized)
    fmt.Println("Uninitialized:", uninitialized)

    var emptyString string        // Will be "" (empty string)
    var defaultBool bool          // Will be false
    var defaultFloat float64      // Will be 0.0
    var defaultPointer *int       // Will be nil
    fmt.Println("Default string:", emptyString == "")
    fmt.Println("Default bool:", defaultBool)
    fmt.Println("Default float:", defaultFloat)
    fmt.Println("Default pointer is nil:", defaultPointer == nil)

    // Example 5: Multiple Variables Declaration
    var a, b, c int = 1, 2, 3
    fmt.Println("Multiple variables:", a, b, c)

    // Example 6: Multiple Variables with Different Types
    var (
        firstName string = "Jane"
        lastName  string = "Smith"
        userAge   int    = 28
        isActive  bool   = true
    )
    fmt.Printf("User: %s %s, Age: %d, Active: %t\n", firstName, lastName, userAge, isActive)

    // Example 7: Shadowing Variables
    x := 10
    fmt.Println("Outer x:", x)
    {
        x := 20 // This is a new variable that shadows the outer x
        fmt.Println("Inner x:", x)
    }
    fmt.Println("Outer x after block:", x) // Still 10

    // Example 8: Constants
    const pi = 3.14159
    const (
        daysInWeek   = 7
        hoursInDay   = 24
        minutesInDay = hoursInDay * 60
    )
    fmt.Println("Pi:", pi)
    fmt.Println("Minutes in a day:", minutesInDay)

    // Example 9: Iota for Enumerated Constants
    const (
        Sunday    = iota // 0
        Monday           // 1
        Tuesday          // 2
        Wednesday        // 3
        Thursday         // 4
        Friday           // 5
        Saturday         // 6
    )
    fmt.Println("Monday is day:", Monday)
    fmt.Println("Friday is day:", Friday)

    // Example 10: Type Conversion
    var integer int = 42
    var float float64 = float64(integer)
    var text string = strconv.Itoa(integer) // Integer to string
    
    fmt.Printf("Integer: %d -> Float: %f -> String: %s\n", integer, float, text)

    // Example 11: Working with different numeric types
    var smallInt int8 = 127 // Max value for int8
    // var invalidInt int8 = 128 // Overflow: would cause a compile error
    fmt.Println("Max int8:", smallInt)

    // Example 12: Scope Rules
    scopeExample()
    
    // Example 13: Complex types
    complexNum := 3 + 4i
    fmt.Printf("Complex number: %v, Real: %f, Imaginary: %f, Magnitude: %f\n", 
        complexNum, real(complexNum), imag(complexNum), 
        math.Sqrt(float64(real(complexNum)*real(complexNum) + imag(complexNum)*imag(complexNum))))
}

func scopeExample() {
    // This demonstrates variable scope
    outerVar := "I'm at function level"
    fmt.Println(outerVar)
    
    if true {
        innerVar := "I'm block-scoped"
        fmt.Println(innerVar)
        fmt.Println(outerVar) // Can access outer variable
    }
    
    // Uncommenting this would cause a compile error - innerVar is out of scope
    // fmt.Println(innerVar)
}

/*
 * Best Practices for Variables in Go:
 *
 * 1. Use short variable declarations (:=) in functions
 * 2. Use meaningful, descriptive names
 * 3. Use camelCase for local variables and PascalCase for exported (public) variables
 * 4. Keep variable scope as limited as possible
 * 5. Choose appropriate types based on data requirements
 * 6. Use constants for values that won't change
 * 7. Avoid unused variables (they cause compile errors)
 * 8. Consider zero values when designing your code
 */

File: real-life/cpp/variable.cpp
#include <iostream>
#include <string>
using namespace std;

/*
 * Variables in C++: A Detailed Explanation
 *
 * A variable is a named storage location in memory that holds a value.
 * Variables allow programs to store, retrieve, and manipulate data.
 *
 * Key Points:
 * 1. Variables have a type, which determines the kind of data they can store.
 * 2. Variables must be declared before use.
 * 3. The value of a variable can be changed during program execution.
 * 4. Variables have a scope, which defines where they can be accessed.
 *
 * Syntax:
 * <data_type> <variable_name> = <initial_value>;
 *
 * Example:
 * int age = 25;
 *
 * Common Data Types in C++:
 * - int: Stores integers (e.g., 1, -5, 100)
 * - float: Stores floating-point numbers (e.g., 3.14, -0.01)
 * - double: Stores double-precision floating-point numbers
 * - char: Stores a single character (e.g., 'A', 'z')
 * - string: Stores a sequence of characters (requires #include <string>)
 * - bool: Stores a boolean value (true or false)
 *
 * Let's explore variables with examples.
 */

int main()
{
    // Example 1: Integer Variable
    int age = 25; // Declaring and initializing an integer variable
    cout << "Age: " << age << endl;

    // Example 2: Floating-Point Variable
    float pi = 3.14f; // Declaring and initializing a float variable
    cout << "Value of Pi: " << pi << endl;

    // Example 3: Double Variable
    double largeNumber = 123456.789; // Declaring and initializing a double variable
    cout << "Large Number: " << largeNumber << endl;

    // Example 4: Character Variable
    char grade = 'A'; // Declaring and initializing a char variable
    cout << "Grade: " << grade << endl;

    // Example 5: String Variable
    string name = "John Doe"; // Declaring and initializing a string variable
    cout << "Name: " << name << endl;

    // Example 6: Boolean Variable
    bool isPassed = true; // Declaring and initializing a boolean variable
    cout << "Passed: " << (isPassed ? "Yes" : "No") << endl;

    // Example 7: Modifying Variable Values
    age = 30; // Changing the value of the variable
    cout << "Updated Age: " << age << endl;

    // Example 8: Scope of Variables
    if (true)
    {
        int localVariable = 100; // This variable is only accessible within this block
        cout << "Local Variable: " << localVariable << endl;
    }
    // Uncommenting the next line will cause an error because localVariable is out of scope
    // cout << localVariable << endl;

    // Example 9: Constants
    const double gravity = 9.8; // Declaring a constant variable
    cout << "Gravity: " << gravity << endl;
    // Uncommenting the next line will cause an error because constants cannot be modified
    // gravity = 10.0;

    return 0;
}

/*
 * Summary:
 * - Variables are essential for storing and manipulating data in a program.
 * - Always choose the appropriate data type for your variables.
 * - Use meaningful names for variables to make your code more readable.
 * - Understand the scope and lifetime of variables to avoid errors.
 * - Use constants for values that should not change during program execution.
 */